### PR TITLE
Add a compile method

### PR DESCRIPTION
**Pull request by jderusse**
This PR adds a `Compile` method to the constraints to optimize constraints computation.

By caching the compiled code, we will be able to perform much more complex optimization on MultiConstraint https://github.com/composer/semver/pull/84#issuecomment-620502945

By using this new feature in the most used called locations in Composer (`Pool`, `PoolBuilder` and `ComposerRepository`) I get 15% CPU https://blackfire.io/profiles/compare/cfa5e43e-0510-4b24-bd44-722dc0c3c7ce/graph

Not covered by the current PR:
- `$compareBranches=true` => adds complexity in Constraint::compile method and called 15x less often. But that could be doables tell me if you think it worth it
- `MutliConstraint->match(MultiConstraint)`

Here is a sample of code generated by the compile method
```php
(version_compare($v, '1.1.0.0', '==')) || (!$b && version_compare($v, '2.0.0.0', '>='))
```

### CODE REVIEW

File: src/CompiledMatcher.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    public function match(ConstraintInterface $constraint, $operator, $version)
+    {
+        $cacheKey = $operator.$constraint;
+        if (!isset($this->checked[$cacheKey])) {
+            $sha = \sha1($cacheKey);
+            $file = $this->cacheDir.'/constraint_'.$sha;
+            $function = 'match_'.$sha;
+            if (!\is_file($file)) {
+                \file_put_contents($file, '<?php function '.$function.'($v, $b){return '.$constraint->compile($operator).';}');
```
Comment (Seldaek) [inline src/CompiledMatcher.php:47]: Is there a big gain in caching to disk? Is it really worth it vs creating lambdas via eval() for example?

File: src/CompiledMatcher.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ */
+class CompiledMatcher
+{
+    private $cacheDir;
+
+    private $checked;
+
+    public function __construct($cacheDir = null)
+    {
+        $this->cacheDir = $cacheDir ? $cacheDir : \sys_get_temp_dir();
```
Comment (Seldaek) [inline src/CompiledMatcher.php:27]: I'd at least use `sys_get_temp_dir().'/composer-constraints'` or something to avoid creating thousands of files in /tmp

File: src/CompiledMatcher.php (hunk 1 above)
Comment (stof) [reply src/CompiledMatcher.php:47]: we should at least keep a mode without `eval`, to account for hosts disabling it.

File: src/CompiledMatcher.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+     *
+     * @return mixed
+     */
+    public function match(ConstraintInterface $constraint, $operator, $version)
+    {
+        $cacheKey = $operator.$constraint;
+        if (!isset($this->checked[$cacheKey])) {
+            $sha = \sha1($cacheKey);
+            $file = $this->cacheDir.'/constraint_'.$sha;
+            $function = 'match_'.$sha;
```
Comment (stof) [inline src/CompiledMatcher.php:45]: if you generate named functions, you should namespace them, or prefix their name, to avoid conflicts with other packages.

File: src/CompiledMatcher.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+     * @param string              $version
+     *
+     * @return mixed
+     */
+    public function match(ConstraintInterface $constraint, $operator, $version)
+    {
+        $cacheKey = $operator.$constraint;
+        if (!isset($this->checked[$cacheKey])) {
+            $sha = \sha1($cacheKey);
+            $file = $this->cacheDir.'/constraint_'.$sha;
```
Comment (stof) [inline src/CompiledMatcher.php:44]: I would add a `.php` extension on the file

Comment (stof) [inline src/CompiledMatcher.php:44]: shouldn't the cache dir or the cache file contain some kind of reference to the version of the compilation logic ? With the current approach, if we fix a bug in the compilation logic, all users of composer will have to delete the temporary cache files to be sure that the old buggy functions are not used.

File: src/Constraint/Constraint.php [hunk 5]
```diff
@@ -171,6 +171,81 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
...
+    public function compile($otherOperator) {
+        $isBranch = 'dev-' === substr($this->version, 0, 4);
+
+        if (self::OP_NE === $this->operator) {
+            if ($isBranch) {
+                if ($otherOperator === self::OP_EQ) {
+                    return 'false';
+                }
+
+                return 'true';
```
Comment (stof) [inline src/Constraint/Constraint.php:183]: This logic looks wrong to me. What if the provided version is the same branch ?

File: src/Constraint/CompilableConstraintInterface.php [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Composer\Semver\Constraint;
+
+interface CompilableConstraintInterface extends ConstraintInterface
+{
+    /**
+     * Provides a compiled version of the constraint for the given operator
+     * Executor of compile version have to provide 2 variables:
+     * - $v = the string version to compare with
+     * - $b = whether or not the version is a branch (starts with "dev-")
```
Comment (stof) [inline src/Constraint/CompilableConstraintInterface.php:20]: starting with `dev-` is about being a non-comparable branch, not just about being a branch. `2.x-dev` is also a branch

File: src/CompiledMatcher.php (hunk 1 above)
Comment (Seldaek) [reply src/CompiledMatcher.php:47]: Hosts without eval can just run without this optimization. Updates should not be run on these hosts anyway. The main question to me is whether writing tons of files helps or not, if it's not worth a significant gain I'd rather avoid touching the filesystem in that way.

File: src/Constraint/MultiConstraint.php [hunk 7]
```diff
@@ -31,6 +31,8 @@ class MultiConstraint implements ConstraintInterface
     /** @var Bound|null */
     protected $upperBound;
 
+    private $cacheString;
```
Comment (stof) [inline src/Constraint/MultiConstraint.php:34]: This cache of `__toString` should be extracted to a separate PR IMO, as it also benefits other optimization work.

File: tests/Constraint/MultiConstraintTest.php [hunk 8]
```diff
@@ -83,6 +84,7 @@ public function testMultiVersionMatchFails()
         $multiRequire = new MultiConstraint(array($this->versionRequireStart, $this->versionRequireEnd));
 
         $this->assertFalse($multiRequire->matches($versionProvide));
+        $this->assertFalse($this->matchCompiled($multiRequire, '==', 1.2));
```
Comment (stof) [inline tests/Constraint/MultiConstraintTest.php:87]: The same kind of assertions should be added in all tests.

File: src/CompiledMatcher.php (hunk 1 above)
Comment (jderusse) [reply src/CompiledMatcher.php:47]: I'm also afraid of writing so many file on disk and understand what you mean.
Just for the record, it's 1 file per evaluated constraint (ie. `^5.0`, `^5.1` are 2 files)
If somebody have tips to avoid writing so much files.

I run composer up on magento project with 3 implementations:
- [composer original](https://blackfire.io/profiles/a74958be-1823-4364-a8a5-5921778fe09f/graph)
- [with file](https://blackfire.io/profiles/d97a1207-65dd-433b-a466-d1c64f8aa076/graph) => [-20%](https://blackfire.io/profiles/compare/0a7269b7-9e1f-48e3-b88c-8ac9ff08f048/graph)
- [with eval](https://blackfire.io/profiles/3298571c-222b-4950-9d69-146fae5b0f2f/graph) => [-1%](https://blackfire.io/profiles/compare/a504a829-47af-4061-bfdc-6ec4869d6a40/graph)

The difference is: Eval requires to recompile constraints every times.

But IMHO **those number are temporary**, this open the field to write more complex optimization on MultiConstraints.

File: src/CompiledMatcher.php (hunk 3 above)
Comment (jderusse) [reply src/CompiledMatcher.php:45]: good catch, fixed (prefixed name of function)

File: src/CompiledMatcher.php (hunk 2 above)
Comment (jderusse) [reply src/CompiledMatcher.php:27]: I prefixed it in cacheKey, but your idea of doing it in constructor is better. 
Fixed

File: src/CompiledMatcher.php (hunk 4 above)
Comment (jderusse) [reply src/CompiledMatcher.php:44]: thanks

Comment (jderusse) [reply src/CompiledMatcher.php:44]: Great idea!.

Do we have an easy way to find the current version of the installed lib?

File: src/Constraint/Constraint.php (hunk 5 above)
Comment (jderusse) [reply src/Constraint/Constraint.php:183]: This is the current behavior: https://github.com/composer/semver/blob/master/src/Constraint/Constraint.php#L193

in fact this path is requireOperator = `!=` and providedOperators is not `==`. see `if ($otherOperator === self::OP_EQ) {` at line 179

matrix is:
```
!dev-master & !1.0
!dev-master & >1.0
!dev-master & <dev-foo
```

File: src/Constraint/CompilableConstraintInterface.php (hunk 6 above)
Comment (jderusse) [reply src/Constraint/CompilableConstraintInterface.php:20]: Comment fixed, thanks

File: tests/Constraint/MultiConstraintTest.php (hunk 8 above)
Comment (jderusse) [reply tests/Constraint/MultiConstraintTest.php:87]: This should be the case. 
I only let non supported matches: `Multi->match(Multi)`. If I missed one please tell me.

File: src/CompiledMatcher.php (hunk 4 above)
Comment (stof) [reply src/CompiledMatcher.php:44]: `Composer\InstalledVersions` in Composer 2 :smile: 

But we might not need to bust the cache on each version. We could have an internal constant storing an "optimization version", that we bump when we change the optimization logic.

**Comment by jderusse** [comment]
Reviewing the `compile` function is very hard. 
I've add a "matrix" test that take all case (combining operators and type of version) and compare the result of `matches` with `compile` to ensure both implementations are identical

Which triggers 4 errors that sounds legit to me
```
==3.0-b2 matches >3.0-beta2 => current = true, mine = false
==3.0-b2 matches <3.0-beta2 => current = true, mine = false
==3.0-beta2 matches >3.0-b2 => current = true, mine = false
==3.0-beta2 matches <3.0-b2 => current = true, mine = false
```

File: src/CompiledMatcher.php (hunk 1 above)
Comment (stof) [reply src/CompiledMatcher.php:48]: @Seldaek this would require making the CompiledMatcher optional, which is not the case currently. And `composer install` also runs the solver. Even if we don't need the optimization in that case, we need the matcher to work.

File: src/Constraint/Constraint.php [hunk 9]
```diff
@@ -171,6 +171,77 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
...
+            }
+
+            return 'true';
+        } elseif (self::OP_EQ === $this->operator) {
+            if ($isBranch) {
+                if (in_array($otherOperator, array(self::OP_LT, self::OP_GT, self::OP_NE))) {
+                    return 'false';
+                }
+
+                return sprintf('$b && $v === \'%s\'', $this->version);
```
Comment (stof) [inline src/Constraint/Constraint.php:197]: use `var_export($this->version, true)` to generate the PHP string representation rather than `sprintf('\'%s\'', $this->version)`. Otherwise, you might face issues in some cases (potentially even allow arbitrary code injection in the composer solver). AFAIK, single quotes are not forbidden in versions when using `dev-*` versions for instance.

File: src/Constraint/MultiConstraint.php [hunk 10]
```diff
@@ -75,6 +77,19 @@ public function isDisjunctive()
         return !$this->conjunctive;
     }
 
+    public function compile($otherOperator)
+    {
+        if (!$this->conjunctive) {
+            return '('.implode(') || (',\array_map(function(ConstraintInterface $constraint) use ($otherOperator) {
+                return $constraint->compile($otherOperator);
```
Comment (stof) [inline src/Constraint/MultiConstraint.php:84]: What if the inner constraints are not compilable ones ?

Comment (jderusse) [reply src/Constraint/MultiConstraint.php:84]: People should not call `compile` on a custom constraint that does not implement the new interface.
I thought about letting PHP throw an exception.

Do you think we should handle such case (ie. Plugin inject Custom Constraint in Composer)?

File: src/CompiledMatcher.php (hunk 1 above)
Comment (Seldaek) [reply src/CompiledMatcher.php:48]: @stof you just need a little more imagination.. this function could `return $constraint->match(new Constraint($op, $version))` if `eval` is not callable. Not such a big deal. Anyway this is obviously solvable one way or another, not worth discussing at this point.

@jderusse is your eval case just evaling the compiled constraint? Or evaling a function if none exists yet, and calling it if it was created already? Because I see 310730 calls to CompiledMatcher and 310730 eval calls in your profile, so it seems to me like you called eval every time, which then really bypasses any optimization potential. Ideally we'd just create an in-memory array of functions instead of dumping them to filesystem. I understand this means the compiled constraints can not be reused across runs, but I'd like to really understand if reusing them is *that* beneficial. I can also try myself when I find a minute though..

File: src/Constraint/MultiConstraint.php (hunk 10 above)
Comment (stof) [reply src/Constraint/MultiConstraint.php:84]: Well, that's exactly what you do here, as you don't check at all whether the inner constraints of the MultiConstraint are custom or no.

And people won't call it. Composer will. If it is possible to have a non-compilable MultiConstraint, we need to have a working way to bailout (not a PHP Fatal Error), so that the matcher can fallback to non-optimized mode. The compilation is just an optimization, so we should not fail hard when we cannot do it.

Comment (Seldaek) [reply src/Constraint/MultiConstraint.php:84]: I am not aware of any custom constraint implementations out in the wild.. Strictly speaking it would be good to check for the interface and fail if it is not compilable, but if this has an adverse effect performance wise we maybe rather just leave it as is and let people doing custom stuff figure out why they get a fatal..

File: src/CompiledMatcher.php (hunk 1 above)
Comment (jderusse) [reply src/CompiledMatcher.php:48]: @Seldaek My eval compiled the code once, then evaluated each time.

```php
        $cacheKey = $operator.$constraint;
        if (!isset($this->checked[$cacheKey])) {
            $code = $this->checked[$cacheKey] = '$r='.$constraint->compile($operator).';';
        } else {
            $code = $this->checked[$cacheKey];
        }

        $v = $version;
        $b = $v[0] === 'd' && 'dev-' === substr($v, 0, 4);

        eval($code);

        return $r;
```

Following your suggestion, I evaluated it once and reuse the function
```php
        $cacheKey = $operator.$constraint;
        if (!isset(static::$checked[$cacheKey])) {
            $sha = \sha1($cacheKey);
            $function = 'composer_semver_constraint_'.$sha;
            static::$checked[$cacheKey] = $function;
            eval('function '.$function.'($v, $b){return '.$constraint->compile($operator).';}');
            $function = '\\'.$function;
        } else {
            $function = static::$checked[$cacheKey];
        }

        $v = $version;

        return $function($version, $v[0] === 'd' && 'dev-' === substr($v, 0, 4));
```
and get better result. The [eval function is almost as fast as](https://blackfire.io/profiles/compare/4d9908a3-ad07-4d43-8994-1fda87b7fb8b/graph) the file one.
Letting me think to use the eval in this PR. 

When we'll improve the MultiConstraint optimize function and the ratio "time saved with compiled"/"time to compile" will decrease we'll can think about using file

File: src/Constraint/Constraint.php (hunk 9 above)
Comment (jderusse) [reply src/Constraint/Constraint.php]: thanks

File: src/Constraint/MultiConstraint.php (hunk 10 above)
Comment (jderusse) [reply src/Constraint/MultiConstraint.php:88]: Fixed (catch the exception and fallback to class constraint match when a constraint is not compilable)

File: src/CompiledMatcher.php (hunk 1 above)
Comment (Seldaek) [reply src/CompiledMatcher.php]: Ok that's great news thanks for testing again. You could probably even skip the sha1 call by doing `static::$checked[$cacheKey] = $function = eval('return function ($v, $b) { return ...; };')` so that would create anonymous functions.

Something like https://3v4l.org/l7YlZ in case it's not clear.

Comment (jderusse) [reply src/CompiledMatcher.php]: Clever !!

File: src/CompiledMatcher.php (hunk 4 above)
Comment (jderusse) [reply src/CompiledMatcher.php]: not needed anymore, switched to `eval` impl

File: src/CompiledMatcher.php [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+use Composer\Semver\Constraint\Constraint;
+use Composer\Semver\Constraint\ConstraintInterface;
+use Composer\Semver\Constraint\NotCompilableConstraintException;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class CompiledMatcher
+{
+    private static $checked = array();
```
Comment (stof) [inline src/CompiledMatcher.php:23]: I would rename this property to `compiledCheckerCache` or something like that.

File: src/CompiledMatcher.php [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        if (self::$enabled === null) {
+            self::$enabled = !in_array('eval', explode(',', ini_get('disable_functions')));
+        }
+        if (!self::$enabled) {
+            return $constraint->matches(new Constraint(self::$transOpInt[$operator], $version));
+        }
+
+        $cacheKey = $operator.$constraint;
+        if (!isset(static::$checked[$cacheKey])) {
```
Comment (stof) [inline src/CompiledMatcher.php:54]: private static stufff **must** be accessed with `self::`, not with `static::`. Otherwise, if you extend the class, the late static binding will try to access `ChildClass::$checked`, which won't work.

File: src/Constraint/CompilableConstraintInterface.php [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+     * Provides a compiled version of the constraint for the given operator
+     * Executor of compile version have to provide 2 variables:
+     * - $v = the string version to compare with
+     * - $b = whether or not the version is a non-comparable branch (starts with "dev-")
+     *
+     * @see Constraint::OP_* for the list of available operators.
+     *
+     * @param int $operator
+     *
+     * @return string
```
Comment (stof) [inline src/Constraint/CompilableConstraintInterface.php:26]: I would add `@throws NotCompilableConstraintException if the constraint cannot be compiled`, as that's part of the API of this method

File: src/Constraint/CompilableConstraintInterface.php [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace Composer\Semver\Constraint;
+
+interface CompilableConstraintInterface extends ConstraintInterface
+{
+    /**
+     * Provides a compiled version of the constraint for the given operator
+     * Executor of compile version have to provide 2 variables:
+     * - $v = the string version to compare with
+     * - $b = whether or not the version is a non-comparable branch (starts with "dev-")
```
Comment (stof) [inline src/Constraint/CompilableConstraintInterface.php:20]: I would document that this compiled code must be a PHP expression (not arbitrary PHP statements)

File: src/CompiledMatcher.php (hunk 12 above)
Comment (staabm) [inline src/CompiledMatcher.php:54]: in php < 8.0 using `\array_key_exists()` is faster then `isset()`

File: src/CompiledMatcher.php [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+            try {
+                $code = $constraint->compile($operator);
```
Comment (staabm) [inline src/CompiledMatcher.php:56]: should this have an additional fast-path with `$constraint instanceof CompilableConstraintInterface`?

File: src/CompiledMatcher.php (hunk 12 above)
Comment (stof) [reply src/CompiledMatcher.php:54]: are you sure about your sentence ? I always heard the opposite. So either you swapped the alternatives, or you swapped the version check.

File: src/CompiledMatcher.php (hunk 15 above)
Comment (stof) [reply src/CompiledMatcher.php:56]: that's not even a fast path. That's simply necessary to be working.

File: src/CompiledMatcher.php (hunk 12 above)
Comment (staabm) [reply src/CompiledMatcher.php:54]: you are right.. mixed it up.

in php <8.0 the usual workaround is
```
if (isset($array[$key]) || \array_key_exists($key, $array)) {
```

while in php8+
```
if (\array_key_exists($key, $array)) {
```
is faster

Comment (jderusse) [reply src/CompiledMatcher.php:54]: From every bench I see, that's the opposite for php <7.2. And since php>7.2 time are ~ equivalent
Note: the old trick replacing `array_key_exists(...)` by `isset(...) || array_key_exists(...)`  are not need anymore since 7.2 as both method have the same performance

File: src/CompiledMatcher.php (hunk 11 above)
Comment (jderusse) [reply src/CompiledMatcher.php]: thanks

File: src/CompiledMatcher.php (hunk 12 above)
Comment (jderusse) [reply src/CompiledMatcher.php]: fixed

File: src/Constraint/CompilableConstraintInterface.php (hunk 14 above)
Comment (jderusse) [reply src/Constraint/CompilableConstraintInterface.php:21]: thanks

File: src/CompiledMatcher.php (hunk 15 above)
Comment (jderusse) [reply src/CompiledMatcher.php:60]: good catch, fixed

File: src/CompiledMatcher.php [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+        $cacheKey = $operator.$constraint;
+        if (!isset(self::$compiledCheckerCache[$cacheKey])) {
+            try {
+                if (!$constraint instanceof CompilableConstraintInterface) {
+                    throw new NotCompilableConstraintException(sprintf('The constraint "%s" is not compilable.', (string) $constraint));
+                }
+                $code = $constraint->compile($operator);
+                self::$compiledCheckerCache[$cacheKey] = $function = eval('return function($v, $b){return '.$code.';};');
+            } catch (NotCompilableConstraintException $e) {
+                self::$compiledCheckerCache[$cacheKey] = $function = static function($v, $b) use ($constraint, $operator) {
```
Comment (Seldaek) [inline src/CompiledMatcher.php:63]: The `static` addition will break PHP 5.3 I am afraid.

File: src/CompiledMatcher.php [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+            self::$enabled = !in_array('eval', explode(',', ini_get('disable_functions')));
+        }
+        if (!self::$enabled) {
+            return $constraint->matches(new Constraint(self::$transOpInt[$operator], $version));
+        }
+
+        $cacheKey = $operator.$constraint;
+        if (!isset(self::$compiledCheckerCache[$cacheKey])) {
+            try {
+                if (!$constraint instanceof CompilableConstraintInterface) {
```
Comment (stof) [inline src/CompiledMatcher.php:57]: I would move it at the beginning, with the `!self::$enabled` check instead. This way, non-compilable constraint don't register a useless entry in the cache (reducing memory usage of that cache for something not bringing any benefit)

File: src/CompiledMatcher.php [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+            } catch (NotCompilableConstraintException $e) {
+                self::$compiledCheckerCache[$cacheKey] = $function = function($v, $b) use ($constraint, $operator) {
+                    return $constraint->matches(new Constraint(CompiledMatcher::$transOpInt[$operator], $v));
+                };
+            }
+        } else {
+            $function = self::$compiledCheckerCache[$cacheKey];
+        }
+
+        $v = $version;
```
Comment (stof) [inline src/CompiledMatcher.php:71]: no need to rename the variable IMO

File: src/CompiledMatcher.php (hunk 17 above)
Comment (jderusse) [reply src/CompiledMatcher.php]: indeed. no need to cache it

File: src/CompiledMatcher.php (hunk 18 above)
Comment (jderusse) [reply src/CompiledMatcher.php]: thanks

File: src/CompiledMatcher.php (hunk 16 above)
Comment (jderusse) [reply src/CompiledMatcher.php]: fixed

**Comment by naderman** [comment]
This looks great to me!

File: src/CompiledMatcher.php [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+        }
+        if (!self::$enabled || !$constraint instanceof CompilableConstraintInterface) {
+            return $constraint->matches(new Constraint(self::$transOpInt[$operator], $version));
+        }
+
+        $cacheKey = $operator.$constraint;
+        if (!isset(self::$compiledCheckerCache[$cacheKey])) {
+            try {
+                $code = $constraint->compile($operator);
+                self::$compiledCheckerCache[$cacheKey] = $function = eval('return function($v, $b){return '.$code.';};');
```
[the remark this answers is not in the document]
Comment (naderman) [reply src/CompiledMatcher.php:58]: Instead of using eval? I mean you can call eval inside the compile() method instead of here, but you're going to need to actually turn the php code string into something executable somewhere? I think it's nice to have the compile method around to access the code directly and potentially write it out elsewhere for other purposes, so this is good as-is?

[the remark this answers is not in the document]
Comment (staabm) [reply src/CompiledMatcher.php:58]: Maybe I am missing something... the only constraint which I can see within this PR which requires the `eval()` is the multi-constraint.. all others dont seem to generate dynamic code from strings, right?

[the remark this answers is not in the document]
Comment (staabm) [reply src/CompiledMatcher.php:58]: Scratch that: the problem is the recursive nature of MultiConstraint->compile()

[the remark this answers is not in the document]
Comment (jderusse) [reply src/CompiledMatcher.php:58]: hmm.. That's a goof idea, but I'm not sure.
From one hand, It'll remove this eval. but from another hand, closures returned by Constraint are not mergeable in MultiConstraint. Which means:
- it will generate a lot of closures: one per constraint + one per multi (even if the same constraint it used by several Multi)
- it would reduce performances, by calling several closures instead of 1 that evaluate the complete rule
- this close the door to future enhancement using the file_system instead of eval.

WDYT?

[the remark this answers is not in the document]
Comment (staabm) [reply src/CompiledMatcher.php:58]: Yeah I realized that it will not work in the full picture :).

At best we merge as is and incremental improve with another PR

[the remark this answers is not in the document]
Comment (jderusse) [reply src/CompiledMatcher.php:58]: I think the idea of @staabm was:
```php
function compile() {
  if (self::OP_NE === $this->operator) {
    if ($isBranch) {
      if ($otherOperator === self::OP_EQ) {
        return fuction() {return false;};
      }
      return fuction() {return true;};
    }
    if ($otherOperator === self::OP_EQ) {
      $version = $this->version;
      return function($b, $v) use ($version) {
        return !$b && version_compare($v, $version, '!=');
    }
  }
  return fuction() {return true;};
  ...
}

```

used:
```php
$fn = $constraint->compile();

$fn($b, $v);
```

[the remark this answers is not in the document]
Comment (staabm) [reply src/CompiledMatcher.php:58]: Exactly

[the remark this answers is not in the document]
Comment (naderman) [reply src/CompiledMatcher.php:58]: @jderusse I'd say, benchmark it? We're definitely working on something that will do more multi/multi comparisons btw https://github.com/composer/composer/pull/8850

[the remark this answers is not in the document]
Comment (jderusse) [reply src/CompiledMatcher.php:58]: Benchmark available [here](https://blackfire.io/profiles/compare/804e3ae2-6c64-4d35-bc09-29b6c0414f38/graph) 
+4% CPU and +1% memory. 

it's not as good as current Impl. but It's not that bad.. 
It remove the "eval" call. But the drawback is we can't persist the compilation result anymore

WDYT 
ping @Toflar @Seldaek

[the remark this answers is not in the document]
Comment (Toflar) [reply src/CompiledMatcher.php:58]: Hmmm...Persisting would have the advantage that it could „learn“ with consecutive calls. If we were to keep a list of precompiled constraints in a global Composer cache for example. Evaling from a stored file can be dangerous though so it would require to be optional maybe? The Symfony ExpressionLanguage can compile as well as evaluate. Maybe semver should provide both and Composer/other libs use what‘s best? Not sure if worth the effort though?

[the remark this answers is not in the document]
Comment (staabm) [reply src/CompiledMatcher.php:58]: > it's not as good as current Impl. but It's not that bad..

Thx for checking. I would leave it as is for now

File: src/Constraint/Constraint.php [hunk 20]
```diff
@@ -171,6 +171,73 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
         return version_compare($a, $b, $operator);
     }
 
+    public function compile($otherOperator) {
+        $otherOperator = self::$transOpStr[$otherOperator];
+
+        $isBranch = 'dev-' === substr($this->version, 0, 4);
```
Comment (staabm) [inline src/Constraint/Constraint.php:177]: This could also expressed using `[]` string offsets in case blackfire thinks substr() is too hot

File: src/Constraint/Constraint.php [hunk 21]
```diff
@@ -171,6 +171,73 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
...
+            }
+
+            if ($otherOperator === self::OP_EQ) {
+                return sprintf('!$b && version_compare($v, %s, \'!=\')', \var_export($this->version, true));
+            }
+
+            return 'true';
+        } elseif (self::OP_EQ === $this->operator) {
+            if ($isBranch) {
+                if (in_array($otherOperator, array(self::OP_LT, self::OP_GT, self::OP_NE))) {
```
Comment (staabm) [inline src/Constraint/Constraint.php:195]: strict in_array everywhere?

File: src/Constraint/Constraint.php [hunk 22]
```diff
@@ -171,6 +171,73 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
...
+
+        if ($otherOperator === self::OP_NE) {
+            return 'true';
+        }
+
+        if ($isBranch) {
+            return 'false';
+        }
+
+        $codeComparison = sprintf('version_compare($v, %s, \'%s\')', \var_export($this->version, true), self::$transOpInt[$this->operator]);
```
Comment (staabm) [inline src/Constraint/Constraint.php:227]: Should the compiled code also be fully qualified?

File: src/CompiledMatcher.php (hunk 19 above)
[the remark this answers is not in the document]
Comment (staabm) [reply src/CompiledMatcher.php:58]: @jderusse could you post a profile of the PR as is?

File: src/CompiledMatcher.php [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+
+use Composer\Semver\Constraint\CompilableConstraintInterface;
+use Composer\Semver\Constraint\Constraint;
+use Composer\Semver\Constraint\ConstraintInterface;
+use Composer\Semver\Constraint\NotCompilableConstraintException;
+
+/**
+ * Helper class to evaluate constraint by compiling and reusing the code to evaluate
+ */
+class CompiledMatcher
```
Comment (staabm) [inline src/CompiledMatcher.php:22]: Should it be `CompilingMatcher`?

File: src/Constraint/Constraint.php (hunk 21 above)
Comment (jderusse) [reply src/Constraint/Constraint.php:195]: fixed

File: src/Constraint/Constraint.php (hunk 22 above)
Comment (jderusse) [reply src/Constraint/Constraint.php:227]: indeed

File: src/Constraint/Constraint.php (hunk 20 above)
Comment (jderusse) [reply src/Constraint/Constraint.php:177]: done

File: src/CompiledMatcher.php (hunk 19 above)
[the remark this answers is not in the document]
Comment (jderusse) [reply src/CompiledMatcher.php]: profile: https://blackfire.io/profiles/8379ae17-f4c4-4b59-83d7-b7c13bc9f115/graph
comparison with original: https://blackfire.io/profiles/compare/cfa5e43e-0510-4b24-bd44-722dc0c3c7ce/graph

File: src/Constraint/MultiConstraint.php [hunk 24]
```diff
@@ -111,7 +111,7 @@ public function matches(ConstraintInterface $provider)
     {
         if (false === $this->conjunctive) {
             foreach ($this->constraints as $constraint) {
-                if ($constraint->matches($provider)) {
+                if ($provider->matches($constraint)) {
```
Comment (Seldaek) [inline src/Constraint/MultiConstraint.php:114]: Not sure what the idea is here, but pretty sure you can't do this change safely.

Comment (jderusse) [reply src/Constraint/MultiConstraint.php:114]: oh, constraint are not swappable? I thought `$a->matche($b) === $b->matches($a)`
The idea were a small optimization:
- `Constraint` is not able to matches `MultiConstraint` and fallback to the provider
- The probability that `$constraint` beiing a `Constraint` is higher than `$provider` beiing a `Constraint`

By swapping the call, it avoid `$constraint` to fallback to `$provider->matches($this)` and avoid 1 method call

If you are not sure about the assumption `$a->matche($b) === $b->matches($a)` I can revert

Comment (Seldaek) [reply src/Constraint/MultiConstraint.php:114]: Well, just have a look at the test suite result :)

File: src/CompilingMatcher.php [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+     * @param string              $version
+     *
+     * @return mixed
+     */
+    public static function match(ConstraintInterface $constraint, $operator, $version)
+    {
+        if (self::$enabled === null) {
+            self::$enabled = !in_array('eval', explode(',', ini_get('disable_functions')));
+        }
+        if (!self::$enabled || !$constraint instanceof CompilableConstraintInterface) {
```
Comment (staabm) [inline src/CompilingMatcher.php:50]: Would it speedup things when doing this instanceof check before calling into CompilingMatcher->match() ?

Comment (stof) [reply src/CompilingMatcher.php:50]: Well, built-in constraint are all compilable. So not calling the CompilingMatcher would be an edge case when a plugin implements the interface by itself.

File: src/Constraint/MultiConstraint.php (hunk 24 above)
Comment (jderusse) [reply src/Constraint/MultiConstraint.php]: reverted

**Comment by Seldaek** [comment]
Tried this and seeing about 20% CPU time reduction on pure dependency solving for ~0.3% memory increase, so totally worth it. Thanks @jderusse and all reviewers for the great work here!
